feature: Native MCP support (#3101)
pydantic/pydantic-ai: docs/builtin-tools.md

print(result.output)
#> Here are some examples of my data: Pen, Paper, Pencil.
```

For OpenAI Responses, you can use a connector by specifying a special `x-openai-connector:` URL:

_(This example is complete, it can be run "as is")_

```py {title="mcp_server_configured_connector_id.py"}
import os

from pydantic_ai import Agent, MCPServerTool

agent = Agent(
    'openai-responses:gpt-4o',
    builtin_tools=[
        MCPServerTool(
            id='your-mcp-server',
            url='x-openai-connector:connector_googlecalendar',
            authorization_token=os.getenv('GITHUB_ACCESS_TOKEN', 'mock-access-token'),
            allowed_tools=['search_repositories', 'list_commits'],
            description='Your MCP Server',
            headers={'X-CUSTOM-HEADER': 'custom-value'},
        )
    ]
)

result = agent.run_sync('Give me some examples of my products.')
print(result.output)
#> Here are some examples of my data: Pen, Paper, Pencil.
```

_(This example is complete, it can be run "as is")_

#### Provider Support

| Parameter             | OpenAI | Anthropic |
|-----------------------|--------|-----------|
| `url`                 | ✅ | ✅ |
| `allowed_tools`       | ✅ | ✅ |

> | `authorization_token` | ✅ | ✅ |

Reviewer: ```suggestion
| `authorization_token` | ✅ | ✅ |
| `allowed_tools`       | ✅ | ✅ |
```

The token seems more common than limiting tools

_(This example is complete, it can be run "as is")_

## MCP Server Tool

The [`MCPServerTool`][pydantic_ai.builtin_tools.MCPServerTool] allows your agent to pass MCP configurations in context,
so that the agent can offload MCP calls and parsing to the provider.

This requires the MCP server to live at a public URL the provider can reach and does not support many of the advanced features of Pydantic AI's agent-side [MCP support](mcp/client.md),
but can result in optimized context use and caching, and faster performance due to the lack of a round-trip back to Pydantic AI.

### Provider Support

| Provider | Supported | Notes                 |
|----------|-----------|-----------------------|

> | OpenAI Responses | ✅ | Full feature support  |

Reviewer: ```suggestion
| OpenAI Responses | ✅ | Full feature support. [Connectors](https://platform.openai.com/docs/guides/tools-connectors-mcp#connectors) can be used by specifying a special `x-openai-connector:<connector_id>` URL. |
```

from pydantic_ai import Agent, MCPServerTool

agent = Agent(
    'openai-responses:gpt-5',
    builtin_tools=[
        MCPServerTool(
            id='your-mcp-server',
            url='https://api.githubcopilot.com/mcp/',
        )
    ]
)

result = agent.run_sync('Give me some examples of my products.')
print(result.output)
#> Here are some examples of my data: Pen, Paper, Pencil.
```

_(This example is complete, it can be run "as is")_

### Configuration Options

The `MCPServerTool` supports several configuration parameters for custom MCP servers:

```py {title="mcp_server_configured_url.py"}
from pydantic_ai import Agent, MCPServerTool

agent = Agent(
    'openai-responses:gpt-5',
    builtin_tools=[
        MCPServerTool(
            id='your-mcp-server',  # required field
            url='https://api.githubcopilot.com/mcp/',  # required field
            allowed_tools=['search_repositories', 'list_commits'],
            description='Your MCP Server',
            headers={'X-CUSTOM-HEADER': 'custom-value'},
        )
    ]
)

result = agent.run_sync('Give me some examples of my products.')
print(result.output)
#> Here are some examples of my data: Pen, Paper, Pencil.
```

> For OpenAI Responses, you can use a connector by specifying a special `x-openai-connector:` URL:

Reviewer: ```suggestion
For OpenAI Responses, you can use a [connector](https://platform.openai.com/docs/guides/tools-connectors-mcp#connectors) by specifying a special `x-openai-connector:` URL:
```

        )
    ]
)

result = agent.run_sync('Give me some examples of my products.')
print(result.output)
#> Here are some examples of my data: Pen, Paper, Pencil.
```

_(This example is complete, it can be run "as is")_

### Configuration Options

The `MCPServerTool` supports several configuration parameters for custom MCP servers:

```py {title="mcp_server_configured_url.py"}
import os

from pydantic_ai import Agent, MCPServerTool

agent = Agent(
    'openai-responses:gpt-4o',
    builtin_tools=[
        MCPServerTool(
            id='your-mcp-server',  # required field
            url='https://api.githubcopilot.com/mcp/',  # optional field, use `url` or `provider_metadata`
            authorization_token=os.getenv('GITHUB_ACCESS_TOKEN', 'mock-access-token'),  # optional field
            allowed_tools=['search_repositories', 'list_commits'],  # optional field
            description='Your MCP Server',  # optional field
            headers={'X-CUSTOM-HEADER': 'custom-value'},  # optional field
        )
    ]
)

result = agent.run_sync('Give me some examples of my products.')
print(result.output)
#> Here are some examples of my data: Pen, Paper, Pencil.
```

> For OpenAI Responses, you can use connector ID instead of URL:

Reviewer: ```suggestion
For OpenAI Responses, you can use a connector by specifying a special `x-openai-connector:` URL:
```

_(This example is complete, it can be run "as is")_

```py {title="mcp_server_configured_connector_id.py"}
import os

from pydantic_ai import Agent, MCPServerTool

agent = Agent(
    'openai-responses:gpt-4o',
    builtin_tools=[
        MCPServerTool(
            id='your-mcp-server',  # required field
            url='x-openai-connector:connector_googlecalendar',  # required field
            authorization_token=os.getenv('GITHUB_ACCESS_TOKEN', 'mock-access-token'),  # optional field
            allowed_tools=['search_repositories', 'list_commits'],  # optional field
            description='Your MCP Server',  # optional field

>             headers={'X-CUSTOM-HEADER': 'custom-value'},  # optional field

Reviewer: We can remove this required/optional comments if we modify the first example to only use the required ones.

        )
    ]
)

result = agent.run_sync('Give me some examples of my products.')
print(result.output)
#> Here are some examples of my data: Pen, Paper, Pencil.
```

_(This example is complete, it can be run "as is")_

#### Provider Support

| Parameter           | OpenAI | Anthropic | Notes                                                                                                               |
|---------------------|--------|-----------|---------------------------------------------------------------------------------------------------------------------|

> | `url`               | ✅ | ✅ | For OpenAI Responses, it is possible to use `connector_id` by providing it as `x-openai-connector:<connector_id>`   |

Reviewer: We can move this to the "Provider Support" table, in Notes. Because now `url` is not a "configuration option" anymore, but a required field with a special value.
